File: Mac/ES_R_DevKit_Mac/MEME_Academic/UserSetting.m
//
//  UserSetting.m
//  MEME_Academic
//
//  Created by Celleus on 2022/09/14.
//  Copyright © 2022 jins-jp. All rights reserved.
//

#import "UserSetting.h"
#import "Const.h"

@implementation UserSetting

//+ (void)clear {
//    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
//    [userDefaults setInteger:0 forKey:kConst_Chart1_Type];
//    [userDefaults setInteger:0 forKey:kConst_Chart2_Type];
//    [userDefaults setInteger:0 forKey:kConst_Chart3_Type];
//    [userDefaults setDouble:0.0 forKey:kConst_X_Axis];
//    [userDefaults setDouble:0.0 forKey:kConst_Y_Axis];
//    [userDefaults setDouble:0.0 forKey:kConst_Z_Axis];
//    [userDefaults setBool:false forKey:kConst_ShowSaveFileDialog];
//    [userDefaults setBool:false forKey:kConst_ExtermalOutputSocket];
//    [userDefaults setObject:nil forKey:kConst_LocalPort];
//    [userDefaults setObject:nil forKey:kConst_LocalAddress];
//}

// =============================================================================
#pragma mark - fristSetting
// =============================================================================

+ (void)fristSetting {
    if ([[NSUserDefaults standardUserDefaults] objectForKey:kConst_LocalPort] == nil) {
        NSLog(@"初期設定開始");
        [self defaultSetting];
    }
    else {
        NSLog(@"初期設定済み");
    }
}

// =============================================================================
#pragma mark - defaultSetting
// =============================================================================

+ (void)defaultSetting {
    [self createDefaultSaveDirectory];
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
//    [userDefaults setInteger:0 forKey:kConst_Chart1_Type];
//    [userDefaults setInteger:0 forKey:kConst_Chart2_Type];
//    [userDefaults setInteger:0 forKey:kConst_Chart3_Type];
    [userDefaults setDouble:0.0 forKey:kConst_X_Axis];
    [userDefaults setDouble:0.0 forKey:kConst_Y_Axis];
    [userDefaults setDouble:0.0 forKey:kConst_Z_Axis];
    [userDefaults setBool:false forKey:kConst_ShowSaveFileDialog];
    [userDefaults setBool:false forKey:kConst_ExtermalOutputSocket];
    [userDefaults setObject:@"88" forKey:kConst_LocalPort];
//    [userDefaults setObject:nil forKey:kConst_LocalAddress];
}

// =============================================================================
#pragma mark - createDefaultSaveDirectory
// =============================================================================
+ (void)createDefaultSaveDirectory {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    NSString *defaultDirectory =  [documentsDirectory stringByAppendingPathComponent:@"/JINS/MEME_Academic"];
    NSLog(@"defaultDirectory:%@",defaultDirectory);
    if (![[NSFileManager defaultManager] isExecutableFileAtPath:defaultDirectory]) {
        NSLog(@"ディレクトリがないので作成");
        if ([[NSFileManager defaultManager] createDirectoryAtPath:defaultDirectory withIntermediateDirectories:YES attributes:nil error:nil]) {
            NSLog(@"ディレクトリ作成 成功");
            NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
            [userDefaults setObject:[NSString stringWithFormat:@"file://%@",defaultDirectory] forKey:kConst_SaveFilePath];
        }
        else {
            NSLog(@"ディレクトリ作成 失敗");
            NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
            [userDefaults setObject:@"" forKey:kConst_SaveFilePath];
        }
    }
    else {
        NSLog(@"既にディレクトリがある");
        NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
        [userDefaults setObject:[NSString stringWithFormat:@"file://%@",defaultDirectory] forKey:kConst_SaveFilePath];
    }
}

// =============================================================================
#pragma mark - Home
// =============================================================================

//+ (void)setChart1Type:(int)value {
//    [[NSUserDefaults standardUserDefaults] setInteger:value forKey:kConst_Chart1_Type];
//}
//+ (long)getChart1Type {
//    return [[NSUserDefaults standardUserDefaults] integerForKey:kConst_Chart1_Type];
//}
//
//+ (void)setChart2Type:(int)value {
//    [[NSUserDefaults standardUserDefaults] setInteger:value forKey:kConst_Chart2_Type];
//}
//+ (long)getChart2Type {
//    return [[NSUserDefaults standardUserDefaults] integerForKey:kConst_Chart2_Type];
//}
//
//+ (void)setChart3Type:(int)value {
//    [[NSUserDefaults standardUserDefaults] setInteger:value forKey:kConst_Chart3_Type];
//}
//+ (long)getChart3Type {
//    return [[NSUserDefaults standardUserDefaults] integerForKey:kConst_Chart3_Type];
//}

// =============================================================================
#pragma mark - Setting
// =============================================================================

+ (void)setSaveFilePath:(id)value {
    [[NSUserDefaults standardUserDefaults] setObject:value forKey:kConst_SaveFilePath];
}
+ (NSString *)getSaveFilePath {
    return [[NSUserDefaults standardUserDefaults] objectForKey:kConst_SaveFilePath];
}

+ (void)setXAxis:(double)value {
    [[NSUserDefaults standardUserDefaults] setDouble:value forKey:kConst_X_Axis];
}
+ (double)getXAxis {
    return [[NSUserDefaults standardUserDefaults] doubleForKey:kConst_X_Axis];
}

+ (void)setYAxis:(double)value {
    [[NSUserDefaults standardUserDefaults] setDouble:value forKey:kConst_Y_Axis];
}
+ (double)getYAxis {
    return [[NSUserDefaults standardUserDefaults] doubleForKey:kConst_Y_Axis];
}

+ (void)setZAxis:(double)value {
    [[NSUserDefaults standardUserDefaults] setDouble:value forKey:kConst_Z_Axis];
}
+ (double)getZAxis {
    return [[NSUserDefaults standardUserDefaults] doubleForKey:kConst_Z_Axis];
}

+ (void)setShowSaveFileDialog:(BOOL)value {
    [[NSUserDefaults standardUserDefaults] setBool:value forKey:kConst_ShowSaveFileDialog];
}
+ (BOOL)getShowSaveFileDialog {
    return [[NSUserDefaults standardUserDefaults] boolForKey:kConst_ShowSaveFileDialog];
}

+ (void)setExtermalOutputSocket:(BOOL)value {
    [[NSUserDefaults standardUserDefaults] setBool:value forKey:kConst_ExtermalOutputSocket];
}
+ (BOOL)getExtermalOutputSocket {
    return [[NSUserDefaults standardUserDefaults] boolForKey:kConst_ExtermalOutputSocket];
}

+ (void)setLocalPort:(id)value {
    [[NSUserDefaults standardUserDefaults] setObject:value forKey:kConst_LocalPort];
}
+ (NSString *)getLocalPort {
    return [[NSUserDefaults standardUserDefaults] objectForKey:kConst_LocalPort];
}

//+ (void)setLocalAddress:(id)value {
//    [[NSUserDefaults standardUserDefaults] setObject:value forKey:kConst_LocalAddress];
//}
//+ (NSString *)getLocalAddress {
//    return [[NSUserDefaults standardUserDefaults] objectForKey:kConst_LocalAddress];
//}

@end

File: Mac/ES_R_DevKit_Mac/MEME_Academic/Utility/UserSetting.swift
//
//  UserSetting.swift
//  MEME_Academic
//
//  Created by Celleus on 2022/09/14.
//  Copyright © 2022 jins-jp. All rights reserved.
//

import Foundation

class UserSetting: NSObject {

    class func fristSetting() {
        if UserDefaults.standard.object(forKey: kConst_LocalPort) == nil {
            NSLog("初期設定開始")
            defaultSetting()
        } else {
            NSLog("初期設定済み")
        }
    }

    class func defaultSetting() {
        createDefaultSaveDirectory()
        let userDefaults = UserDefaults.standard
        userDefaults.set(0.0, forKey: kConst_X_Axis)
        userDefaults.set(0.0, forKey: kConst_Y_Axis)
        userDefaults.set(0.0, forKey: kConst_Z_Axis)
        userDefaults.set(false, forKey: kConst_ShowSaveFileDialog)
        userDefaults.set(false, forKey: kConst_ExtermalOutputSocket)
        userDefaults.set("88", forKey: kConst_LocalPort)
        userDefaults.set(true, forKey: kConst_ConvertToLocalTime)
    }

    private class func createDefaultSaveDirectory() {
        let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)
        guard let documentsDirectory = paths.first else { return }
        let defaultDirectory = (documentsDirectory as NSString).appendingPathComponent("/JINS/MEME_Academic")
        NSLog("defaultDirectory:%@", defaultDirectory)
        let userDefaults = UserDefaults.standard
        if !FileManager.default.isExecutableFile(atPath: defaultDirectory) {
            NSLog("ディレクトリがないので作成")
            do {
                try FileManager.default.createDirectory(atPath: defaultDirectory, withIntermediateDirectories: true, attributes: nil)
                NSLog("ディレクトリ作成 成功")
                userDefaults.set("file://\(defaultDirectory)", forKey: kConst_SaveFilePath)
            } catch {
                NSLog("ディレクトリ作成 失敗")
                userDefaults.set("", forKey: kConst_SaveFilePath)
            }
        } else {
            NSLog("既にディレクトリがある")
            userDefaults.set("file://\(defaultDirectory)", forKey: kConst_SaveFilePath)
        }
    }

    // MARK: - Setting

    class func setSaveFilePath(_ value: Any?) {
        UserDefaults.standard.set(value, forKey: kConst_SaveFilePath)
    }
    class func getSaveFilePath() -> String {
        return UserDefaults.standard.object(forKey: kConst_SaveFilePath) as? String ?? ""
    }

    class func setXAxis(_ value: Double) {
        UserDefaults.standard.set(value, forKey: kConst_X_Axis)
    }
    class func getXAxis() -> Double {
        return UserDefaults.standard.double(forKey: kConst_X_Axis)
    }

    class func setYAxis(_ value: Double) {
        UserDefaults.standard.set(value, forKey: kConst_Y_Axis)
    }
    class func getYAxis() -> Double {
        return UserDefaults.standard.double(forKey: kConst_Y_Axis)
    }

    class func setZAxis(_ value: Double) {
        UserDefaults.standard.set(value, forKey: kConst_Z_Axis)
    }
    class func getZAxis() -> Double {
        return UserDefaults.standard.double(forKey: kConst_Z_Axis)
    }

    class func setShowSaveFileDialog(_ value: Bool) {
        UserDefaults.standard.set(value, forKey: kConst_ShowSaveFileDialog)
    }
    class func getShowSaveFileDialog() -> Bool {
        return UserDefaults.standard.bool(forKey: kConst_ShowSaveFileDialog)
    }

    class func setExtermalOutputSocket(_ value: Bool) {
        UserDefaults.standard.set(value, forKey: kConst_ExtermalOutputSocket)
    }
    class func getExtermalOutputSocket() -> Bool {
        return UserDefaults.standard.bool(forKey: kConst_ExtermalOutputSocket)
    }

    /// 時刻表示をローカルタイムへ変換するか（既定：ON）。
    /// CSV／ソケットへ記録する時刻は常にUTCで、この設定は表示（チャートのX軸ラベル）にのみ効く。
    class func setConvertToLocalTime(_ value: Bool) {
        UserDefaults.standard.set(value, forKey: kConst_ConvertToLocalTime)
    }
    class func getConvertToLocalTime() -> Bool {
        // 未設定（この設定より前から使っているユーザー）は ON 扱いにする。
        // bool(forKey:) は未設定でも false を返すため、object(forKey:) で有無を判定する。
        guard let value = UserDefaults.standard.object(forKey: kConst_ConvertToLocalTime) as? Bool else {
            return true
        }
        return value
    }

    class func setLocalPort(_ value: Any?) {
        UserDefaults.standard.set(value, forKey: kConst_LocalPort)
    }
    class func getLocalPort() -> String {
        return UserDefaults.standard.object(forKey: kConst_LocalPort) as? String ?? ""
    }
}
